lib/Frontend/ArgsToFrontendOutputsConverter.cpp

```diff
@@ -202,3 +218,225 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+      defaultSupplementaryOutputPathExcludingExtension);
+  if (!dependenciesFilePath)
+    return None;
+
+  auto referenceDependenciesFilePath = determineSupplementaryOutputFilename(
+      OPT_emit_reference_dependencies,
+      pathsFromArguments.ReferenceDependenciesFilePath, "swiftdeps", "",
+      defaultSupplementaryOutputPathExcludingExtension);
+  if (!referenceDependenciesFilePath)
+    return None;
+
+  auto serializedDiagnosticsPath = determineSupplementaryOutputFilename(
+      OPT_serialize_diagnostics, pathsFromArguments.SerializedDiagnosticsPath,
+      "dia", "", defaultSupplementaryOutputPathExcludingExtension);
+
+  if (!serializedDiagnosticsPath)
+    return None;
+
+  auto objcHeaderOutputPath = determineSupplementaryOutputFilename(
+      OPT_emit_objc_header, pathsFromArguments.ObjCHeaderOutputPath, "h", "",
+      defaultSupplementaryOutputPathExcludingExtension);
+  if (!objcHeaderOutputPath)
+    return None;
+
+  auto loadedModuleTracePath = determineSupplementaryOutputFilename(
+      OPT_emit_loaded_module_trace, pathsFromArguments.LoadedModuleTracePath,
+      "trace.json", "", defaultSupplementaryOutputPathExcludingExtension);
+  if (!loadedModuleTracePath)
+    return None;
+
+  auto tbdPath = determineSupplementaryOutputFilename(
+      OPT_emit_tbd, pathsFromArguments.TBDPath, "tbd", "",
+      defaultSupplementaryOutputPathExcludingExtension);
+  if (!tbdPath)
+    return None;
+
+  auto moduleDocOutputPath = determineSupplementaryOutputFilename(
+      OPT_emit_module_doc, pathsFromArguments.ModuleDocOutputPath,
+      SERIALIZED_MODULE_DOC_EXTENSION, "",
+      defaultSupplementaryOutputPathExcludingExtension);
+  if (!moduleDocOutputPath)
+    return None;
+
+  ID emitModuleOption;
+  std::string moduleExtension;
+  std::string mainOutputIfUsableForModule;
+  deriveModulePathParameters(emitModuleOption, moduleExtension,
+                             mainOutputIfUsableForModule);
+
+  auto moduleOutputPath = determineSupplementaryOutputFilename(
+      emitModuleOption, pathsFromArguments.ModuleOutputPath, moduleExtension,
+      mainOutputIfUsableForModule,
+      defaultSupplementaryOutputPathExcludingExtension);
+  if (!moduleOutputPath)
+    return None;
+
+  SupplementaryOutputPaths sop;
+  sop.ObjCHeaderOutputPath = *objcHeaderOutputPath;
+  sop.ModuleOutputPath = *moduleOutputPath;
+  sop.ModuleDocOutputPath = *moduleDocOutputPath;
+  sop.DependenciesFilePath = *dependenciesFilePath;
+  sop.ReferenceDependenciesFilePath = *referenceDependenciesFilePath;
+  sop.SerializedDiagnosticsPath = *serializedDiagnosticsPath;
+  sop.LoadedModuleTracePath = *loadedModuleTracePath;
+  sop.TBDPath = *tbdPath;
+  return sop;
+}
+
+StringRef SupplementaryOutputPathsComputer::
+    deriveDefaultSupplementaryOutputPathExcludingExtension(
+        StringRef outputFilename, const InputFile &input) const {
+  // Put the serialized diagnostics file next to the output file if possible.
```

Comment:
Another specific-rather-than-generic comment.

Reply:
Thank you for the catch! Fixed.